ethpm/manifest.py

```diff
@@ -0,0 +1,171 @@
+            - `website`
+            - `documentation`
+            - `repository`
+        """
+        self.meta = {key: value for key, value in kwargs.items() if value}
+
+    def pretty(self) -> str:
+        """
+        Returns a string of a pretty printed version of this manifest.
+        """
+        attr_dict = self._generate_manifest_attributes()
+        return json.dumps(attr_dict, indent=4, sort_keys=True)
+
+    def minified(self) -> str:
+        """
+        Returns a string of a minified version of this manifest.
+        """
+        attr_dict = self._generate_manifest_attributes()
+        return json.dumps(attr_dict, separators=(",", ":"), sort_keys=True)
+
+    # TODO implement `write_pretty_to_disk` w/o indenting ABI
+    def write_minified_to_disk(self) -> Path:
+        """
+        Returns a path leading to a minified version of this manifest written to disk.
+        Manifest is written to the same dir in which the `contracts/` dir is located.
+        """
+        manifest_contents = self.minified()
+        path = self.solc_path.parent / str(self.version + ".json")
+        path.write_text(manifest_contents)
+        return path
+
+    @to_dict
+    def _generate_manifest_attributes(self) -> Generator[Tuple[str, Any], Any, None]:
+        for key, value in self.__dict__.items():
+            if key in MANIFEST_FIELDS and value:
+                yield key, value
+
+
+@to_dict
+def generate_all_contract_types(
+    solc: Dict[str, Any], contract_types: List[str]
+) -> Generator[Tuple[str, Any], None, None]:
+    """
+    Return a dict representing all of a Manifest's "contract_types" data.
+    """
+    for contract in solc["contracts"]:
+        _, contract_name = contract.split(":")
+        if contract_name in contract_types:
+            yield contract_name, generate_single_contract_type(contract, solc)
+
+
+@to_dict
+def generate_single_contract_type(
+    contract: str, solc: Dict[str, Any]
+) -> Generator[Tuple[str, Any], None, None]:
+    """
+    Return a dict represnting a single contract type.
+    """
+    yield "abi", json.loads(solc["contracts"][contract]["abi"])
+    yield "natspec", json.loads(solc["contracts"][contract]["devdoc"])
+
+
+@to_dict
+def generate_all_sources(
+    solc: Dict[str, Any], solc_path: str, ipfs_backend: BaseIPFSBackend
+) -> Generator[Tuple[str, str], None, None]:
+    """
+    Return a dict containing all of a Manifest's "sources" data.
+    """
+    for contract in solc["contracts"]:
+        contract_path, _ = contract.split(":")
+        source_data = pin_single_source(contract_path, solc_path, ipfs_backend)
+        yield "./{0}".format(contract_path), source_data
+
+
+def pin_single_source(
+    contract_path: str, solc_path: str, ipfs_backend: BaseIPFSBackend
+) -> str:
+    """
+    Return an IPFS URI, after pinning a source file to IPFS via `ipfs_backend`.
+    """
+    [ipfs_return_data] = ipfs_backend.pin_assets(Path(solc_path).parent / contract_path)
```

Comment:
This syntax confused me as I've not seen using list for unpacking return values.  Would it bother you to change this to tuple form?

```python
(x,) = ...
```

This is absolutely a nitpicky thing...

Reply:
In a non-nitpicky thread....

For an MVP, nothing here needs to change, however, I *think* that it would be good for us to think about the lifecycle for manifests.  For packaging, it is normally split into two distinct steps.  `build -> publish`.  This however results in partial publication as part of what I perceive to be the "build" step.  We should explore how we can do a full build, and push this publishing step out till the point where the publish the manifest itself.  Simply writing up an issue for this would be adequate for now so that we don't lose track of it.